### PR TITLE
Add whitepaper experiments

### PR DESCRIPTION
**Pull request by andrei-stoian-zama**
cc @kcelia 

Closes https://github.com/zama-ai/concrete-ml-internal/issues/4380
Closes https://github.com/zama-ai/concrete-ml-internal/issues/4297

### CODE REVIEW

**Review by RomanBredehoft — DISMISSED** [review]

**Review by RomanBredehoft — DISMISSED** [review]

**Review by bcm-at-zama — COMMENTED** [review]
Thanks @kcelia / @andrei-stoian-zama .

Some comments:
- in the intro, maybe I would clarify that the WP had been done with another unpublished compiler
- Wether
- involves quantizes
- "    # The model is compiled through 'compile_torch_model' method": maybe the moment to insist on 6b rounding + approximate PBS + p_error
- is it possible to "print" the model or show its ONNX somehow, such that we see the topology, please (in particular, count the number of PBSs)
- "Timing in FHE: 62.281s per sample.": I would precise it's not simulate here
- BENCH_HPC7A = {20: [0.987, 0.959, 0.995], 50: [0.9745, 0.9477, 3.03]}: what about moving it at the end, just before use, to precise it's execution times measured on another machine

File: use_case_examples/white_paper_experiment/WhitePaperExperiments.ipynb [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Running NN-20 on a 128-core machine machine:Accuracy in fp32 : 98.067% for the test set\n",
+      "Accuracy with FHE-simulation mode : 14.163% for the test set\n",
```
Comment (jfrery) [inline use_case_examples/white_paper_experiment/WhitePaperExperiments.ipynb:515]: Something is wrong with the accuracy here

File: use_case_examples/white_paper_experiment/WhitePaperExperiments.ipynb [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+       "<pandas.io.formats.style.Styler at 0x7f5c34121f40>"
+      ]
+     },
+     "execution_count": 8,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# This benchmark was done on a hpc7 machine with 192-cores\n",
```
Comment (jfrery) [inline use_case_examples/white_paper_experiment/WhitePaperExperiments.ipynb:716]: The notebook deserve a conclusion I think.

**Review by jfrery — CHANGES_REQUESTED** [review]
I probably did my review a bit early as it looks like things are note yet clean. Please make this a draft PR while it's unfinished.

File: use_case_examples/white_paper_experiment/WhitePaperExperiments.ipynb [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Downloading http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz\n",
+      "Failed to download (trying next):\n",
+      "HTTP Error 403: Forbidden\n",
```
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply use_case_examples/white_paper_experiment/WhitePaperExperiments.ipynb:183]: it's like that every time, Lecun must have taken down his mirror of mnist

File: src/concrete/ml/quantization/post_training.py [hunk 4]
```diff
@@ -872,7 +872,7 @@ def _process_layer(
         """
 
         return self._calibrate_layers_activation(
-            True, quantized_op, *calibration_data, quantizers=quantizers
+            False, quantized_op, *calibration_data, quantizers=quantizers
```
Comment (fd0r) [inline src/concrete/ml/quantization/post_training.py:875]: So we accept that we change that?

Comment (andrei-stoian-zama) [reply src/concrete/ml/quantization/post_training.py:875]: I don't see any other solution

File: src/concrete/ml/quantization/post_training.py [hunk 5]
```diff
@@ -387,16 +384,8 @@ def _calibrate_layers_activation(
...
-        # raw operation output (RawOpOutput) data should not be quantized
-        if calibrate_quantized and not isinstance(quant_result, RawOpOutput):
-            assert isinstance(quant_result, QuantizedArray)
-            return (
-                quant_result.dequant(),
-                quant_result.quantizer if isinstance(quant_result, QuantizedArray) else None,
-            )
 
-        # For QAT, the calibration is performed on raw data, performing
+        # For both PTQ and QAT, the calibration is performed on raw data, performing
```
Comment (jfrery) [inline src/concrete/ml/quantization/post_training.py:388]: Can you link to an issue here. We have to track why we did this change somewhere.

File: src/concrete/ml/quantization/post_training.py (hunk 4 above)
Comment (jfrery) [reply src/concrete/ml/quantization/post_training.py]: I reverted that change. Not needed with clipping based fix.

File: src/concrete/ml/quantization/post_training.py (hunk 5 above)
Comment (jfrery) [reply src/concrete/ml/quantization/post_training.py]: reverted

**Review by jfrery — APPROVED** [review]

File: src/concrete/ml/quantization/quantized_ops.py [hunk 6]
```diff
@@ -1612,6 +1612,49 @@ class QuantizedBatchNormalization(QuantizedOp):
...
+        )
+
+        raw_result = self.call_impl(*prepared_inputs, **self.attrs)
+        if isinstance(raw_result, RawOpOutput):
+            return raw_result
+
+        # Check if batch normalization is applied per channel
+        scale = self.constant_inputs[self._params_name_to_input_idx["scale"]].values
+        bias = self.constant_inputs[self._params_name_to_input_idx["bias"]].values
+        if scale.size > 1 or bias.size > 1:
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_ops.py:1638]: there is no better way then this to check if bn is done per channel ?

**Review by RomanBredehoft — DISMISSED** [review]
thanks ! if there's no better way to check if the BN is done per channel or not, then fine for me

just one thing : I'm not sure why this PR is closing https://github.com/zama-ai/concrete-ml-internal/issues/857, is this expected ?

File: src/concrete/ml/quantization/quantized_ops.py (hunk 6 above)
Comment (RomanBredehoft) [reply src/concrete/ml/quantization/quantized_ops.py]: and yes I guess htat means we should add some new tests

**Review by jfrery — APPROVED** [review]

**Review by fd0r — APPROVED** [review]

**Pull request merged**